### PR TITLE
Fix floating numbers parse and respect culture

### PR DESCRIPTION
**Pull request by aeleftheriadis**
Fix StandardFilters methods that parses object to string, that fixes floating point numbers with current that has different decimal separator such as comma.

### CODE REVIEW

**Comment by microalps** [comment]
Please create corresponding tests for all the changed methods.

All the best,

Moshe Alfih
microalps web services

File: src/DotLiquid.Tests/StandardFilterTests.cs [hunk 1]
```diff
@@ -1675,6 +1713,30 @@ public void TestAbs()
...
+
+            Helper.AssertTemplateResult(
+                expected: "17",
+                template: "{{ -17 | abs }}");
+            Helper.AssertTemplateResult(
+                expected: "17",
+                template: "{{ 17 | abs }}");
+            Helper.AssertTemplateResult(
+                expected: "4",
+                template: "{{ 4 | abs }}");
```
Comment (microalps) [inline src/DotLiquid.Tests/StandardFilterTests.cs:1737]: missing culture, what do these tests validate? And no example of fraction in this set of tests.

File: src/DotLiquid.Tests/StandardFilterTests.cs [hunk 2]
```diff
@@ -1699,19 +1761,42 @@ public void TestAtLeast()
         }
 
         [Test]
-        public void TestAtMost()
```
Comment (microalps) [inline src/DotLiquid.Tests/StandardFilterTests.cs:1702]: do not remove existing tests unless there is a good justification for that.

File: src/DotLiquid.Tests/StandardFilterTests.cs [hunk 3]
```diff
@@ -1389,6 +1394,23 @@ public void TestCeil()
...
+            using (CultureHelper.SetCulture("el-GR"))
+            {
+                Helper.AssertTemplateResult("2", "{{ 1,2 | ceil }}");
+                Helper.AssertTemplateResult("2", "{{ 2,0 | ceil }}");
+                Helper.AssertTemplateResult("184", "{{ 183,357 | ceil }}");
+                Helper.AssertTemplateResult("4", "{{ \"3,5\" | ceil }}");
+
+                Assert.Null(StandardFilters.Ceil(_contextV20, ""));
+                Assert.Null(StandardFilters.Ceil(_contextV20, "two"));
+                Assert.AreEqual(4, StandardFilters.Ceil(_contextGreekCultureV20, 3.5));
```
Comment (microalps) [inline src/DotLiquid.Tests/StandardFilterTests.cs:1410]: should this also have a test for passing a string as "3,5"?

**Review by microalps — CHANGES_REQUESTED** [review]

**Comment by aeleftheriadis** [comment]
@microalps  I have updated the tests as reviewed

**Review by aeleftheriadis — COMMENTED** [review]
Reviewed as requested

**Comment by microalps** [comment]
I will look into this shortly. Sorry I missed this!

File: src/DotLiquid.Tests/StandardFilterTests.cs [hunk 4]
```diff
@@ -1389,6 +1394,23 @@ public void TestCeil()
...
+
+        [Test]
+        public void TestCeil_DifferentCultureDecimalSeparator()
+        {
+            using (CultureHelper.SetCulture("el-GR"))
+            {
+                Helper.AssertTemplateResult("2", "{{ 1,2 | ceil }}");
+                Helper.AssertTemplateResult("2", "{{ 2,0 | ceil }}");
+                Helper.AssertTemplateResult("184", "{{ 183,357 | ceil }}");
+                Helper.AssertTemplateResult("4", "{{ \"3,5\" | ceil }}");
```
Comment (microalps) [inline src/DotLiquid.Tests/StandardFilterTests.cs:1406]: I would suggest an additional test for thousands separators for both cultures. e.g. 1.234.567,89 should parse for el-GR and 1,234,567.89 for en.

**Review by microalps — CHANGES_REQUESTED** [review]

**Comment by microalps** [comment]
TestFloor_DifferentCultureDecimalSeparator and TestCeil_DifferentCultureDecimalSeparator do not fail without your fixes. There is missing test for intended scenario(s).

File: src/DotLiquid.Tests/StandardFilterTests.cs [hunk 5]
```diff
@@ -1389,6 +1394,23 @@ public void TestCeil()
...
+
+                Assert.Null(StandardFilters.Ceil(_contextV20, ""));
+                Assert.Null(StandardFilters.Ceil(_contextV20, "two"));
+                Assert.AreEqual(4, StandardFilters.Ceil(_contextGreekCultureV20, "3,5"));
```
Comment (microalps) [inline src/DotLiquid.Tests/StandardFilterTests.cs:1410]: this is within a using and is the reason the test doesn't fail it seems

**Review by microalps — CHANGES_REQUESTED** [review]
I think this covers the tests required

**Review by microalps — CHANGES_REQUESTED** [review]
I think these are the changes needed...

**Comment by microalps** [comment]
Another note: `Currency` method also calls ToString and presumably is also affected

**Comment by microalps** [comment]
The more I think about this, it seems that the library incorrectly expects that each individual filter handles the culture setting. In my opinion, it would be more consistent if we simply changed the threads culture in `RenderInternal` if different than the current setting. The only side-effect would be the removal of the proprietary `{% param culture='en-US' %}` or change its behavior to set during initialization and not during render.

**Comment by microalps** [comment]
Did some more research into this - a fix would probably not be backwards compatible. In testing, I found that `TestEuroCurrencyFromString` has various cultures with decimal separated fractions where those cultures should be comma separated. I assume that the same is true if anyone is relying on passing American numbers as strings with foreign cultures in the context/renderparameters. I'm baffled how to continue here.  I created #555 to track my work thus far on this issue.

**Comment by microalps** [comment]
Closing this as #555 incorporates this patch and some.

**Comment by daviburg** [comment]
> Did some more research into this - a fix would probably not be backwards compatible. In testing, I found that `TestEuroCurrencyFromString` has various cultures with decimal separated fractions where those cultures should be comma separated. I assume that the same is true if anyone is relying on passing American numbers as strings with foreign cultures in the context/renderparameters. I'm baffled how to continue here. I created #555 to track my work thus far on this issue.

Correct, backward compatibility would be a big issue for such type of change. We could only do this if we find a way to have it opt-in so current consumers don't suddenly get different outputs upon updating the version with no code change on their end.

**Pull request closed without merging**